policy_agent/policy_agent.py

```diff
@@ -0,0 +1,643 @@
+            try:
+                arg = ports_obj[try_arg]
+            except KeyError:
+                arg = None
+            return arg
+
+        self.association_list = dict()
+        for subset in self.subsets:
+
+            # Generate a list of new svc profiles per port spec
+            port_rules = list()
+            for port_spec in subset["ports"]:
+                dst_port = verify_args(port_spec, "port")
+                protocol = verify_args(port_spec, "protocol").lower()
+                if protocol and dst_port:
+                    port_rules.append(Rule(action="allow",
+                                           dst_ports=[dst_port],
+                                           protocol=protocol))
+
+            # Create Profile for the subset
+            # TODO: better profile names
+            profile_name = "svc_%s_%s_port_%s_protocol_%s" % (
+                self.namespace, self.name, dst_port, protocol)
+            _log.info("Adding Profile %s to Store" % profile_name)
+
+            if not _datastore_client.profile_exists(profile_name):
+                _log.info("Creating Profile %s in Store" % profile_name)
+                profile_path = PROFILE_PATH % {"profile_id": profile_name}
+                _datastore_client.etcd_client.write(
+                    profile_path + "tags", '["%s"]' % profile_name)
+
+                # Determine rule set
+                default_allow = Rule(action="allow")
+                rules = Rules(id=profile_name,
+                              inbound_rules=port_rules,
+                              outbound_rules=[default_allow])
+
+                # Write rules to profile
+                _datastore_client.etcd_client.write(
+                    profile_path + "rules", rules.to_json())
+
+            else:
+                _log.warning("Profile %s already in Store" % profile_name)
+
+            # Generate list of pod names
+            pods = list()
+            for pod in subset["addresses"]:
+                pods.append(pod["targetRef"]["name"])
+
+            # Assumes profile does not exist
+            self.association_list[profile_name] = pods
+
+        _log.info("association list:\n%s" % self.association_list)
+        return self.association_list
+
+
+def _keep_watch(queue, resource, resource_version):
+    """
+    Called by watcher threads. Adds watch events to Queue
+    """
+    while True:
+        try:
+            response = _get_api_stream(resource, resource_version)
+            for line in response.iter_lines():
+                if line:
+                    queue.put(line)
+        except:
+            # If we hit an exception attempting to watch this path, log it, and retry the watch
+            # after a short sleep in order to prevent tight-looping.  We catch all BaseExceptions
+            # so that the thread never dies.
+            _log.exception("Exception watching path %s", resource)
+            time.sleep(10)
+
+
+def _get_api_token():
```

Comment:
More common function between the plugin and the policy agent.   Don't want to duplicate code, so this should be moved to a central location accessible by both.